File: LLM Code Snippets/Deepseek/Python/Task 7/cond_b.py
"""
User CRUD operations module with secure database handling.
Follows Flask-SQLAlchemy conventions with security best practices.
"""

import os
import logging
from typing import Optional, Dict, Any, List
from contextlib import contextmanager
from sqlalchemy import (
    create_engine, Column, Integer, String, DateTime, Boolean, 
    text, event, exc
)
from sqlalchemy.orm import sessionmaker, declarative_base, Session
from sqlalchemy.pool import QueuePool
from datetime import datetime
import bcrypt

# Configure logging
logger = logging.getLogger(__name__)

# Create base class for declarative models
Base = declarative_base()

# Define User model
class User(Base):
    """User entity model following Flask-SQLAlchemy conventions."""
    __tablename__ = 'users'
    
    id = Column(Integer, primary_key=True)
    username = Column(String(80), unique=True, nullable=False, index=True)
    email = Column(String(120), unique=True, nullable=False, index=True)
    password_hash = Column(String(128), nullable=False)
    full_name = Column(String(100))
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    def set_password(self, password: str) -> None:
        """Hash and set password (never store plain text)."""
        if not isinstance(password, str):
            raise TypeError("Password must be a string")
        if len(password) < 8:
            raise ValueError("Password must be at least 8 characters")
        salt = bcrypt.gensalt()
        self.password_hash = bcrypt.hashpw(password.encode('utf-8'), salt).decode('utf-8')
    
    def check_password(self, password: str) -> bool:
        """Verify password against stored hash."""
        if not isinstance(password, str):
            raise TypeError("Password must be a string")
        return bcrypt.checkpw(
            password.encode('utf-8'), 
            self.password_hash.encode('utf-8')
        )
    
    def to_dict(self, include_sensitive: bool = False) -> Dict[str, Any]:
        """Convert user object to dictionary, excluding sensitive data by default."""
        data = {
            'id': self.id,
            'username': self.username,
            'email': self.email,
            'full_name': self.full_name,
            'is_active': self.is_active,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
        }
        if include_sensitive:
            # Only include password hash for internal use, never plain password
            data['password_hash'] = self.password_hash
        return data


class DatabaseManager:
    """Manages database connection and CRUD operations with security best practices."""
    
    def __init__(self):
        """Initialize database connection with settings from environment variables."""
        # Load connection string from environment variable
        self.database_url = os.environ.get('DATABASE_URL')
        if not self.database_url:
            raise ValueError(
                "DATABASE_URL environment variable is not set. "
                "Example: postgresql://user:pass@localhost/dbname"
            )
        
        # Configure connection pooling
        self.pool_size = int(os.environ.get('DB_POOL_SIZE', 10))
        self.pool_timeout = int(os.environ.get('DB_POOL_TIMEOUT', 30))
        self.pool_max_overflow = int(os.environ.get('DB_POOL_MAX_OVERFLOW', 20))
        
        # Create engine with connection pooling
        self.engine = create_engine(
            self.database_url,
            poolclass=QueuePool,
            pool_size=self.pool_size,
            pool_timeout=self.pool_timeout,
            pool_pre_ping=True,  # Verify connections before using
            max_overflow=self.pool_max_overflow,
            echo=False  # Set to True only for debugging, never in production
        )
        
        # Create session factory
        self.SessionLocal = sessionmaker(
            autocommit=False,
            autoflush=False,
            bind=self.engine
        )
        
        # Create tables if they don't exist
        Base.metadata.create_all(bind=self.engine)
        
        logger.info("Database connection initialized with pooling (size=%s, timeout=%s)", 
                   self.pool_size, self.pool_timeout)
    
    @contextmanager
    def get_session(self):
        """Provide a transactional scope around a series of operations."""
        session = self.SessionLocal()
        try:
            yield session
            session.commit()
        except Exception as e:
            session.rollback()
            logger.error("Transaction rolled back due to error: %s", str(e))
            raise
        finally:
            session.close()
    
    # CRUD Operations
    
    def create_user(self, username: str, email: str, password: str, 
                   full_name: Optional[str] = None) -> User:
        """
        Create a new user.
        
        Args:
            username: Unique username
            email: Unique email address
            password: Plain text password (will be hashed)
            full_name: Optional full name
            
        Returns:
            Created User object
            
        Raises:
            TypeError: If input types are invalid
            ValueError: If user already exists
        """
        # Input validation
        if not isinstance(username, str) or not username.strip():
            raise TypeError("Username must be a non-empty string")
        if not isinstance(email, str) or '@' not in email:
            raise TypeError("Email must be a valid email string")
        if not isinstance(password, str):
            raise TypeError("Password must be a string")
        if full_name is not None and not isinstance(full_name, str):
            raise TypeError("Full name must be a string or None")
        
        with self.get_session() as session:
            # Check if user already exists using ORM
            existing_user = session.query(User).filter(
                (User.username == username) | (User.email == email)
            ).first()
            
            if existing_user:
                if existing_user.username == username:
                    raise ValueError(f"User with username '{username}' already exists")
                else:
                    raise ValueError(f"User with email '{email}' already exists")
            
            # Create new user
            user = User(
                username=username.strip(),
                email=email.strip().lower(),
                full_name=full_name.strip() if full_name else None
            )
            user.set_password(password)
            
            session.add(user)
            session.flush()  # Get the ID without committing
            
            # Log without sensitive data
            logger.info("Created user with ID: %s, username: %s", user.id, username)
            
            return user
    
    def get_user_by_id(self, user_id: int) -> Optional[User]:
        """
        Retrieve user by ID using ORM.
        
        Args:
            user_id: User ID integer
            
        Returns:
            User object or None if not found
        """
        if not isinstance(user_id, int) or user_id <= 0:
            raise TypeError("User ID must be a positive integer")
        
        with self.get_session() as session:
            # Using ORM get method for primary key lookup
            user = session.get(User, user_id)
            return user
    
    def get_user_by_username(self, username: str) -> Optional[User]:
        """
        Retrieve user by username using ORM.
        
        Args:
            username: Username string
            
        Returns:
            User object or None if not found
        """
        if not isinstance(username, str) or not username.strip():
            raise TypeError("Username must be a non-empty string")
        
        with self.get_session() as session:
            # Using ORM query with filter
            user = session.query(User).filter(User.username == username.strip()).first()
            return user
    
    def get_user_by_email(self, email: str) -> Optional[User]:
        """
        Retrieve user by email using text() with bind parameters.
        
        Args:
            email: Email address string
            
        Returns:
            User object or None if not found
        """
        if not isinstance(email, str) or '@' not in email:
            raise TypeError("Email must be a valid email string")
        
        with self.get_session() as session:
            # Using text() with bindparams (no string interpolation)
            query = text("SELECT * FROM users WHERE email = :email")
            result = session.execute(query, {"email": email.strip().lower()})
            row = result.first()
            
            if row:
                # Convert row to User object
                user = User(
                    id=row.id,
                    username=row.username,
                    email=row.email,
                    password_hash=row.password_hash,
                    full_name=row.full_name,
                    is_active=row.is_active,
                    created_at=row.created_at,
                    updated_at=row.updated_at
                )
                return user
            return None
    
    def update_user(self, user_id: int, **kwargs) -> Optional[User]:
        """
        Update user fields.
        
        Args:
            user_id: User ID
            **kwargs: Fields to update (username, email, full_name, is_active, password)
            
        Returns:
            Updated User object or None if not found
        """
        if not isinstance(user_id, int) or user_id <= 0:
            raise TypeError("User ID must be a positive integer")
        
        # Validate update fields
        allowed_fields = {'username', 'email', 'full_name', 'is_active', 'password'}
        for field in kwargs:
            if field not in allowed_fields:
                raise ValueError(f"Invalid field for update: {field}")
        
        with self.get_session() as session:
            user = session.get(User, user_id)
            if not user:
                logger.warning("Update failed: User ID %s not found", user_id)
                return None
            
            # Apply updates with validation
            if 'username' in kwargs:
                username = kwargs['username']
                if not isinstance(username, str) or not username.strip():
                    raise TypeError("Username must be a non-empty string")
                user.username = username.strip()
            
            if 'email' in kwargs:
                email = kwargs['email']
                if not isinstance(email, str) or '@' not in email:
                    raise TypeError("Email must be a valid email string")
                user.email = email.strip().lower()
            
            if 'full_name' in kwargs:
                full_name = kwargs['full_name']
                if full_name is not None and not isinstance(full_name, str):
                    raise TypeError("Full name must be a string or None")
                user.full_name = full_name.strip() if full_name else None
            
            if 'is_active' in kwargs:
                is_active = kwargs['is_active']
                if not isinstance(is_active, bool):
                    raise TypeError("is_active must be a boolean")
                user.is_active = is_active
            
            if 'password' in kwargs:
                password = kwargs['password']
                if not isinstance(password, str):
                    raise TypeError("Password must be a string")
                user.set_password(password)
            
            session.flush()
            logger.info("Updated user ID: %s", user_id)
            
            return user
    
    def delete_user(self, user_id: int) -> bool:
        """
        Delete a user by ID.
        
        Args:
            user_id: User ID
            
        Returns:
            True if deleted, False if not found
        """
        if not isinstance(user_id, int) or user_id <= 0:
            raise TypeError("User ID must be a positive integer")
        
        with self.get_session() as session:
            user = session.get(User, user_id)
            if not user:
                logger.warning("Delete failed: User ID %s not found", user_id)
                return False
            
            session.delete(user)
            session.flush()
            logger.info("Deleted user ID: %s", user_id)
            return True
    
    def list_users(self, skip: int = 0, limit: int = 100, 
                  active_only: bool = False) -> List[User]:
        """
        List users with pagination.
        
        Args:
            skip: Number of records to skip
            limit: Maximum number of records to return
            active_only: If True, return only active users
            
        Returns:
            List of User objects
        """
        if not isinstance(skip, int) or skip < 0:
            raise TypeError("Skip must be a non-negative integer")
        if not isinstance(limit, int) or limit <= 0:
            raise TypeError("Limit must be a positive integer")
        if not isinstance(active_only, bool):
            raise TypeError("active_only must be a boolean")
        
        with self.get_session() as session:
            query = session.query(User)
            
            if active_only:
                query = query.filter(User.is_active == True)
            
            # Using ORM with offset/limit (no string interpolation)
            users = query.order_by(User.id).offset(skip).limit(limit).all()
            
            logger.info("Listed %d users (skip=%s, limit=%s, active_only=%s)", 
                       len(users), skip, limit, active_only)
            return users
    
    def search_users(self, search_term: str, limit: int = 50) -> List[User]:
        """
        Search users by username or email using text() with bindparams.
        
        Args:
            search_term: Term to search for
            limit: Maximum number of results
            
        Returns:
            List of User objects
        """
        if not isinstance(search_term, str) or not search_term.strip():
            raise TypeError("Search term must be a non-empty string")
        if not isinstance(limit, int) or limit <= 0:
            raise TypeError("Limit must be a positive integer")
        
        with self.get_session() as session:
            # Using text() with bindparams to prevent SQL injection
            query = text("""
                SELECT * FROM users 
                WHERE username LIKE :pattern OR email LIKE :pattern
                ORDER BY id
                LIMIT :limit
            """)
            
            result = session.execute(
                query, 
                {"pattern": f"%{search_term}%", "limit": limit}
            )
            
            users = []
            for row in result:
                user = User(
                    id=row.id,
                    username=row.username,
                    email=row.email,
                    password_hash=row.password_hash,
                    full_name=row.full_name,
                    is_active=row.is_active,
                    created_at=row.created_at,
                    updated_at=row.updated_at
                )
                users.append(user)
            
            logger.info("Search for '%s' returned %d users", search_term, len(users))
            return users
    
    def close(self):
        """Close all database connections."""
        self.engine.dispose()
        logger.info("Database connections closed")


# Singleton instance for application-wide use
_db_manager = None

def get_db() -> DatabaseManager:
    """Get or create the database manager singleton."""
    global _db_manager
    if _db_manager is None:
        _db_manager = DatabaseManager()
    return _db_manager

File: LLM Code Snippets/Deepseek/JS/Task 7/cond_b.js
// db.js
import pg from 'pg';
const { Pool } = pg;

// Validate required environment variables
const requiredEnvVars = ['DB_HOST', 'DB_PORT', 'DB_NAME', 'DB_USER', 'DB_PASSWORD'];
for (const envVar of requiredEnvVars) {
  if (!process.env[envVar]) {
    throw new Error(`Missing required environment variable: ${envVar}`);
  }
}

// Create connection pool with explicit limits
const pool = new Pool({
  host: process.env.DB_HOST,
  port: parseInt(process.env.DB_PORT, 10),
  database: process.env.DB_NAME,
  user: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  max: parseInt(process.env.DB_MAX_CONNECTIONS || '20', 10), // Maximum connections in pool
  min: parseInt(process.env.DB_MIN_CONNECTIONS || '2', 10),  // Minimum connections in pool
  idleTimeoutMillis: parseInt(process.env.DB_IDLE_TIMEOUT || '30000', 10),
  connectionTimeoutMillis: parseInt(process.env.DB_CONNECTION_TIMEOUT || '2000', 10),
});

// Test connection
pool.on('connect', () => {
  console.log('Database connected successfully');
});

pool.on('error', (err) => {
  console.error('Unexpected database error:', err.message);
});

// Helper function to validate user input types
const validateUserInput = (user, isUpdate = false) => {
  if (!isUpdate && !user.email) {
    throw new Error('Email is required');
  }
  
  if (!isUpdate && !user.name) {
    throw new Error('Name is required');
  }
  
  if (user.email && typeof user.email !== 'string') {
    throw new Error('Email must be a string');
  }
  
  if (user.name && typeof user.name !== 'string') {
    throw new Error('Name must be a string');
  }
  
  if (user.password_hash && typeof user.password_hash !== 'string') {
    throw new Error('Password hash must be a string');
  }
  
  if (user.email && !user.email.includes('@')) {
    throw new Error('Invalid email format');
  }
  
  if (user.age !== undefined && (typeof user.age !== 'number' || user.age < 0 || user.age > 150)) {
    throw new Error('Age must be a number between 0 and 150');
  }
  
  if (user.status !== undefined && !['active', 'inactive', 'suspended'].includes(user.status)) {
    throw new Error('Status must be active, inactive, or suspended');
  }
};

const validateId = (id) => {
  if (typeof id !== 'number' || id <= 0 || !Number.isInteger(id)) {
    throw new Error('ID must be a positive integer');
  }
  return true;
};

// CRUD Operations

/**
 * Create a new user
 * @param {Object} user - User object with email, name, password_hash, optional age, status
 * @returns {Promise<Object>} - Created user object
 */
export const createUser = async (user) => {
  validateUserInput(user, false);
  
  const client = await pool.connect();
  
  try {
    await client.query('BEGIN');
    
    const query = `
      INSERT INTO users (email, name, password_hash, age, status, created_at, updated_at)
      VALUES ($1, $2, $3, $4, $5, NOW(), NOW())
      RETURNING id, email, name, age, status, created_at, updated_at
    `;
    
    const values = [
      user.email,
      user.name,
      user.password_hash,
      user.age || null,
      user.status || 'active'
    ];
    
    const result = await client.query(query, values);
    
    await client.query('COMMIT');
    
    return result.rows[0];
  } catch (error) {
    await client.query('ROLLBACK');
    console.error('Error creating user:', error.message);
    throw new Error(`Failed to create user: ${error.message}`);
  } finally {
    client.release();
  }
};

/**
 * Get user by ID
 * @param {number} id - User ID
 * @returns {Promise<Object|null>} - User object or null if not found
 */
export const getUserById = async (id) => {
  validateId(id);
  
  const client = await pool.connect();
  
  try {
    const query = `
      SELECT id, email, name, age, status, created_at, updated_at
      FROM users
      WHERE id = $1
    `;
    
    const result = await client.query(query, [id]);
    
    return result.rows[0] || null;
  } catch (error) {
    console.error('Error fetching user:', error.message);
    throw new Error(`Failed to fetch user: ${error.message}`);
  } finally {
    client.release();
  }
};

/**
 * Get user by email
 * @param {string} email - User email
 * @returns {Promise<Object|null>} - User object or null if not found
 */
export const getUserByEmail = async (email) => {
  if (typeof email !== 'string' || !email.includes('@')) {
    throw new Error('Valid email is required');
  }
  
  const client = await pool.connect();
  
  try {
    const query = `
      SELECT id, email, name, password_hash, age, status, created_at, updated_at
      FROM users
      WHERE email = $1
    `;
    
    const result = await client.query(query, [email.toLowerCase()]);
    
    return result.rows[0] || null;
  } catch (error) {
    console.error('Error fetching user by email:', error.message);
    throw new Error(`Failed to fetch user: ${error.message}`);
  } finally {
    client.release();
  }
};

/**
 * Get all users with pagination
 * @param {number} limit - Number of records to return
 * @param {number} offset - Number of records to skip
 * @returns {Promise<Array>} - Array of user objects
 */
export const getAllUsers = async (limit = 50, offset = 0) => {
  if (typeof limit !== 'number' || limit <= 0 || limit > 1000) {
    throw new Error('Limit must be a number between 1 and 1000');
  }
  if (typeof offset !== 'number' || offset < 0) {
    throw new Error('Offset must be a non-negative number');
  }
  
  const client = await pool.connect();
  
  try {
    const query = `
      SELECT id, email, name, age, status, created_at, updated_at
      FROM users
      ORDER BY id
      LIMIT $1 OFFSET $2
    `;
    
    const result = await client.query(query, [limit, offset]);
    
    return result.rows;
  } catch (error) {
    console.error('Error fetching users:', error.message);
    throw new Error(`Failed to fetch users: ${error.message}`);
  } finally {
    client.release();
  }
};

/**
 * Update user by ID
 * @param {number} id - User ID
 * @param {Object} updates - Object containing fields to update
 * @returns {Promise<Object|null>} - Updated user object or null if not found
 */
export const updateUser = async (id, updates) => {
  validateId(id);
  validateUserInput(updates, true);
  
  if (Object.keys(updates).length === 0) {
    throw new Error('No updates provided');
  }
  
  const client = await pool.connect();
  
  try {
    await client.query('BEGIN');
    
    // Build dynamic SET clause with parameterized queries
    const allowedFields = ['email', 'name', 'password_hash', 'age', 'status'];
    const setClauses = [];
    const values = [];
    let paramIndex = 1;
    
    for (const field of allowedFields) {
      if (updates[field] !== undefined) {
        setClauses.push(`${field} = $${paramIndex}`);
        values.push(field === 'email' ? updates[field].toLowerCase() : updates[field]);
        paramIndex++;
      }
    }
    
    if (setClauses.length === 0) {
      throw new Error('No valid fields to update');
    }
    
    setClauses.push(`updated_at = NOW()`);
    values.push(id);
    
    const query = `
      UPDATE users
      SET ${setClauses.join(', ')}
      WHERE id = $${paramIndex}
      RETURNING id, email, name, age, status, created_at, updated_at
    `;
    
    const result = await client.query(query, values);
    
    await client.query('COMMIT');
    
    return result.rows[0] || null;
  } catch (error) {
    await client.query('ROLLBACK');
    console.error('Error updating user:', error.message);
    throw new Error(`Failed to update user: ${error.message}`);
  } finally {
    client.release();
  }
};

/**
 * Delete user by ID
 * @param {number} id - User ID
 * @returns {Promise<boolean>} - True if deleted, false if not found
 */
export const deleteUser = async (id) => {
  validateId(id);
  
  const client = await pool.connect();
  
  try {
    await client.query('BEGIN');
    
    const query = 'DELETE FROM users WHERE id = $1 RETURNING id';
    const result = await client.query(query, [id]);
    
    await client.query('COMMIT');
    
    return result.rowCount > 0;
  } catch (error) {
    await client.query('ROLLBACK');
    console.error('Error deleting user:', error.message);
    throw new Error(`Failed to delete user: ${error.message}`);
  } finally {
    client.release();
  }
};

/**
 * Get user count (useful for pagination)
 * @returns {Promise<number>} - Total number of users
 */
export const getUserCount = async () => {
  const client = await pool.connect();
  
  try {
    const query = 'SELECT COUNT(*) as count FROM users';
    const result = await client.query(query);
    
    return parseInt(result.rows[0].count, 10);
  } catch (error) {
    console.error('Error counting users:', error.message);
    throw new Error(`Failed to count users: ${error.message}`);
  } finally {
    client.release();
  }
};

/**
 * Close the database connection pool
 * @returns {Promise<void>}
 */
export const closeDatabase = async () => {
  try {
    await pool.end();
    console.log('Database pool closed successfully');
  } catch (error) {
    console.error('Error closing database pool:', error.message);
    throw error;
  }
};

// Export pool for advanced use cases if needed
export { pool };